[Bug]: Error calling _is_collection on a closed HDF5 dataset

### What happened?

> got the same on a fresh PR while testing against development hdmf etc: https://github.com/dandi/dandi-cli/actions/runs/22961153984/job/66651851285?pr=1816 

 _Originally posted by @yarikoptic in [#1423](https://github.com/hdmf-dev/hdmf/issues/1423#issuecomment-4043649550)_

### Steps to Reproduce

```python
Call `repr(nwbfile)` or `_is_collection` on a closed HDF5 dataset
```

### Traceback

```pytb
________________________________ test_pynwb_io _________________________________
dandi/tests/test_pynwb_utils.py:20: in test_pynwb_io
    assert repr(nwbfile)
           ^^^^^^^^^^^^^
/opt/hostedtoolcache/Python/3.13.12/x64/lib/python3.13/site-packages/hdmf/container.py:629: in __repr__
    if _is_collection(v):
       ^^^^^^^^^^^^^^^^^
/opt/hostedtoolcache/Python/3.13.12/x64/lib/python3.13/site-packages/hdmf/utils.py:870: in _is_collection
    if hasattr(data, "ndim"):
       ^^^^^^^^^^^^^^^^^^^^^
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
/opt/hostedtoolcache/Python/3.13.12/x64/lib/python3.13/site-packages/h5py/_hl/dataset.py:542: in ndim
    return self.id.rank
           ^^^^^^^^^^^^
h5py/h5d.pyx:213: in h5py.h5d.DatasetID.rank.__get__
    ???
h5py/h5d.pyx:214: in h5py.h5d.DatasetID.rank.__get__
    ???
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
h5py/h5d.pyx:372: in h5py.h5d.DatasetID.get_space
    ???
E   RuntimeError: Unable to synchronously get dataspace (identifier is not of specified type)
```

### Operating System

Linux

### Python Version

3.13

### Package Versions

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@
 - Added `_unwrap_scalar` helper in `hdmf.utils` to convert 0-d ndarrays to numpy scalars via `.item()`. This fixes `isinstance` checks against Python scalar types that fail on 0-d ndarrays returned by array-API-conforming libraries (e.g., zarr v3 scalar indexing). @h-mayorquin [#1415](https://github.com/hdmf-dev/hdmf/pull/1415)
 
 ### Fixed
+- Fixed `_is_collection` raising `RuntimeError` when accessing `ndim` on closed h5py datasets, which caused `__repr__` to crash on containers read from closed HDF5 files. @rly [#1426](https://github.com/hdmf-dev/hdmf/pull/1426)
 - Fixed writing compound dtype datasets with a single field. @rly [#1420](https://github.com/hdmf-dev/hdmf/pull/1420)
 - Replaced undeclared `pyyaml` dependency with `ruamel.yaml` (a declared core dependency) in `test_common_io.py`. The test relied on PyYAML being transitively installed by pip's `h5py`, which is not the case in conda environments. @rly [#1418](https://github.com/hdmf-dev/hdmf/pull/1418)
 

--- a/src/hdmf/utils.py
+++ b/src/hdmf/utils.py
@@ -867,9 +867,16 @@ def _is_collection(data):
     """
     if isinstance(data, (str, bytes)):
         return False
-    if hasattr(data, "ndim"):
-        return data.ndim > 0
-    return hasattr(data, "__len__")
+    try:
+        ndim = data.ndim
+        return ndim > 0
+    except AttributeError:
+        # No ndim attribute (e.g. list, tuple, dict). Fall back to __len__.
+        return hasattr(data, "__len__")
+    except Exception:
+        # Accessing ndim on a closed h5py dataset raises RuntimeError.
+        # Treat inaccessible data as non-collection.
+        return False
 
 
 def _get_length(data) -> int:

--- a/tests/unit/utils_test/test_utils.py
+++ b/tests/unit/utils_test/test_utils.py
@@ -70,6 +70,23 @@ class TestIsCollection(TestCase):
             shape = ()
         self.assertFalse(_is_collection(FakeScalar()))
 
+    def test_ndim_raises_runtime_error(self):
+        """Simulate closed h5py dataset where accessing ndim raises RuntimeError."""
+        class ClosedDataset:
+            @property
+            def ndim(self):
+                raise RuntimeError("File is closed")
+        self.assertFalse(_is_collection(ClosedDataset()))
+
+    def test_closed_h5py_dataset(self):
+        """Test that a real closed h5py dataset is handled gracefully."""
+        import tempfile
+        with tempfile.NamedTemporaryFile(suffix=".h5") as f:
+            hfile = h5py.File(f.name, "w")
+            ds = hfile.create_dataset("data", data=[1, 2, 3])
+            hfile.close()
+            self.assertFalse(_is_collection(ds))
+
 
 class TestGetLength(TestCase):
     """Tests for _get_length helper that gets first-dimension length."""

--- a/tests/unit/utils_test/test_utils.py
+++ b/tests/unit/utils_test/test_utils.py
@@ -6,6 +6,7 @@ from hdmf.container import Data
 from hdmf.data_utils import DataChunkIterator, DataIO
 from hdmf.testing import TestCase
 from hdmf.utils import get_data_shape, to_uint_array, is_newer_version, _is_collection, _get_length, _unwrap_scalar
+from tests.unit.helpers.utils import get_temp_filepath
 
 
 class TestIsCollection(TestCase):
@@ -80,12 +81,15 @@ class TestIsCollection(TestCase):
 
     def test_closed_h5py_dataset(self):
         """Test that a real closed h5py dataset is handled gracefully."""
-        import tempfile
-        with tempfile.NamedTemporaryFile(suffix=".h5") as f:
-            hfile = h5py.File(f.name, "w")
+        path = get_temp_filepath()
+        try:
+            hfile = h5py.File(path, "w")
             ds = hfile.create_dataset("data", data=[1, 2, 3])
             hfile.close()
             self.assertFalse(_is_collection(ds))
+        finally:
+            if os.path.exists(path):
+                os.remove(path)
 
 
 class TestGetLength(TestCase):

